src/task4feedback/simulator/rl/models/a2c.py

```diff
@@ -48,106 +67,103 @@ def compute_returns(self, next_value, rewards):
-            assert len(self.lst_log_probs) == self.step_for_optim 
-            assert len(self.lst_values) == self.step_for_optim 
-            assert len(self.lst_rewards) == self.step_for_optim 
+            assert len(self.log_probs) == self.step_for_optim
+            assert len(self.values) == self.step_for_optim
+            assert len(self.rewards) == self.step_for_optim
 
-            # TODO(hc): replace it with the actual next state
-            next_x, next_gcn_x, next_gcn_edgeindex = \
-                self.construct_next_state(x, gcn_x, gcn_edgeindex)
             # To perform TD to optimize the model, get a state value
-            # of the expected next state from the critic network
+            # of the expected next state from the critic netowrk
             _, next_value = self.a2c_model(
                 NetworkInput(next_x, False, next_gcn_x, next_gcn_edgeindex))
             cat_log_probs = torch.cat(
-                [lp.unsqueeze(0) for lp in self.lst_log_probs])
-            cat_values = torch.cat([v for v in self.lst_values])
-            cat_rewards = torch.cat([r for r in self.lst_rewards])
+                [lp.unsqueeze(0) for lp in self.log_probs])
+            cat_values = torch.cat(self.values)
+            cat_rewards = torch.cat(self.rewards).to(self.device)
             returns = self.compute_returns(next_value, cat_rewards)
-            returns = torch.cat(returns).detach()
-            print("\t log probs:", cat_log_probs)
-            print("\t values:", cat_values)
-            print("\t rewards:", cat_rewards)
-            print("\t return:", returns)
-            print("\t next value:", next_value)
-            print("\t next x:", next_x)
-
+            returns = torch.cat(returns).detach() 
             advantage = returns - cat_values
-            print("\t advantage:", advantage)
-
-            actor_loss = -(cat_log_probs * advantage.detach())
-            print("\t actor loss:", actor_loss)
-            actor_loss = actor_loss.mean()
-            print("\t actor loss mean:", actor_loss)
-
-            critic_loss = advantage.pow(2).mean()
-            print("\t critic loss:", critic_loss)
-
-            loss = actor_loss + 0.5 * critic_loss - 0.001 * self.entropy
-            print("\t loss:", loss)
-            self.steps = 0
-            self.entropy = 0
-            self.lst_log_probs = []
-            self.lst_values = []
-            self.lst_rewards = []
-
+            actor_loss = -(cat_log_probs * advantage.detach()).mean()
+            #critic_loss = advantage.pow(2).mean()
+            critic_loss = 1 * F.mse_loss(cat_values.unsqueeze(-1), returns.unsqueeze(-1))
+            loss = actor_loss + 0.5 * critic_loss - 0.001 * self.entropy_sum
             self.optimizer.zero_grad()
             loss.backward()
+            # torchviz.make_dot(loss, params=dict(self.a2c_model.named_parameters())).render("attacehd", format="png")
             for param in self.a2c_model.parameters():
```

Comment:
Clip grad norm is safer to preserve direction, right? Or am I missing something?

Reply:
this is for gradient exploding. I used this for the past model since I saw that due to high state spaces. But it might not be necessary in here since we all normalize states.

Reply:
i didn't put much energy to this. (https://jamesmccaffrey.wordpress.com/2022/10/17/the-difference-between-pytorch-clip_grad_value_-and-clip_grad_norm_-functions/) just used simple and easy to use as the article says. 

> Do most gradients fall in (-1,1)? 
Input gradients were not within (-1, 1). In the past simulator and model, gradients were aggressive and probs was exploded. I don't have concrete numbers now. But it seems like it is worth to explore in the new model + simulator.